src/radical/entk/tools/darshan.py

```diff
@@ -0,0 +1,173 @@
+            raise RuntimeError('Darshan root directory should be set with '
+                               'either env variable or an absolute path '
+                               f'(provided path: {darshan_runtime_root})')
+        _darshan_runtime_root = darshan_runtime_root or '$DARSHAN_RUNTIME_ROOT'
+
+    global _darshan_activation_cmds
+    global _darshan_env
+
+    if _darshan_activation_cmds is None:
+
+        _darshan_activation_cmds = []
+        for module in modules or []:
+            _darshan_activation_cmds.append(f'module load {module}')
+        for k, v in (env or {}).items():
+            _darshan_activation_cmds.append(f'export {k.upper()}="{v}"')
+
+        _darshan_env = ru.env_prep(pre_exec_cached=_darshan_activation_cmds)
+
+
+# ------------------------------------------------------------------------------
+# decorator to enable darshan for function that generates Pipeline, Stage, Task
+def with_darshan(func,
+                 darshan_runtime_root: Optional[str] = None,
+                 modules: Optional[List[str]] = None,
+                 env: Optional[Dict[str, str]] = None):
+    def wrapper(*args, **kwargs):
+        return enable_darshan(func(*args, **kwargs),
+                              darshan_runtime_root=darshan_runtime_root,
+                              modules=modules,
+                              env=env)
+    return wrapper
+
+
+# ------------------------------------------------------------------------------
+#
+def enable_darshan(pst: Union[Pipeline, Stage, Task, List[Pipeline]],
+                   darshan_runtime_root: Optional[str] = None,
+                   modules: Optional[List[str]] = None,
+                   env: Optional[Dict[str, str]] = None
+                   ) -> Union[Pipeline, Stage, Task]:
+
+    if not pst:
+        raise ValueError('PST object is not provided')
+    elif isinstance(pst, list):
+        if not isinstance(pst[0], Pipeline):
+            raise TypeError('List of Pipelines is not provided')
+    elif not isinstance(pst, (Pipeline, Stage, Task)):
+        raise TypeError('Non PST object provided')
+
+    cache_darshan_env(darshan_runtime_root, modules, env)
+
+    def _enable_darshan(src_task: Task):
+
+        if not src_task.executable:
+            return
+        elif 'libdarshan.so' in src_task.executable:
+            # Darshan is already enabled
+            return
+
+        darshan_log_dir = DARSHAN_LOG_DIR % {'sandbox': '${RP_TASK_SANDBOX}'}
+        darshan_enable  = (f'LD_PRELOAD="{_darshan_runtime_root}'
+                           '/lib/libdarshan.so" ')
+
+        if src_task.cpu_reqs.cpu_processes == 1:
+            darshan_enable += 'DARSHAN_ENABLE_NONMPI=1 '
+
+        src_task.executable  = darshan_enable + src_task.executable
+        src_task.pre_launch += [f'mkdir -p {darshan_log_dir}']
+        src_task.pre_exec.extend(
+            _darshan_activation_cmds +
+            [f'export DARSHAN_LOG_DIR_PATH={darshan_log_dir}'])
```

Comment:
This could (should?) got to the task environment?

Reply:
yes, thus each task would have its own path to logs